src/buffer.cc

```diff
@@ -0,0 +1,83 @@
+#include <micro/core/buffer.h>
+#include <micro/core/message.h>
+
+#include <deque>
+#include <stdexcept>
+
+namespace micro
+{
+    class Buffer::Impl
+    {
+    public:
+        std::deque<std::unique_ptr<Message>> queue;
+        std::size_t max_size;
+
+        Impl(const std::size_t size);
+
+        void push(std::unique_ptr<Message> msg);
+        std::unique_ptr<Message> pop();
+    };
+}
+
+// Implementation
+micro::Buffer::Impl::Impl(const std::size_t size)
+    : queue(std::deque<std::unique_ptr<Message>>())
```

Comment:
This not needed. Here you are copy (or moving in the best case) the deque. Simply remove this line

Reply:
you're totally right! 41a566913c4cf417a08384d195c1ffea780a86ba